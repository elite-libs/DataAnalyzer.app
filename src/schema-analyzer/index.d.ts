// export type TypeDescriptorName = 'enum' | 'nullable' | 'unique';
// export type TypeNameString = '$ref' |
//     'Unknown' |
//     'ObjectId' |
//     'UUID' |
//     'Boolean' |
//     'Date' |
//     'Timestamp' |
//     'Currency' |
//     'Float' |
//     'Number' |
//     'Email' |
//     'String' |
//     'Array' |
//     'Object' |
//     'Null'

// export interface ISchemaAnalyzerOptions {
//     onProgress?: progressCallback | undefined;
//     enumMinimumRowCount?: number | undefined;
//     enumAbsoluteLimit?: number | undefined;
//     enumPercentThreshold?: number | undefined;
//     nullableRowsThreshold?: number | undefined;
//     uniqueRowsThreshold?: number | undefined;
//     strictMatching?: boolean | undefined;
//     disableNestedTypes?: boolean | undefined;
//     bogusSizeThreshold?: number | undefined,
// }

// /**
//  * Includes the results of main top-level schema.
//  */
// export type TypeSummary = {
//     fields: {
//         [x: string]: FieldInfo;
//     };
//     totalRows: number;
//     nestedTypes?: {
//         [x: string]: TypeSummary;
//     } | undefined;
// };
// /**
//  * Describes one or more potential types discovered for a field. The `types` object will have a `$ref` key if any nested structures were found.
//  */
// export type FieldInfo = {
//     /**
//      * - field stats organized by type
//      */
//     types: {
//         [x: TypeNameString]: FieldTypeSummary;
//     };
//     /**
//      * - is the field nullable
//      */
//     nullable: boolean;
//     /**
//      * - is the field unique
//      */
//     unique: boolean;
//     /**
//      * - enumeration detected, the values are listed on this property.
//      */
//     enum?: string[] | number[] | undefined;
// };
// /**
//  * Contains stats for a given field's (potential) type.
//  *
//  * TODO: Add string property for the type name.
//  *    We currently uses object key structure: {"String": FieldTypeSummary}
//  */
// export type FieldTypeSummary = {
//     /**
//      * - for nested type support.
//      */
//     typeAlias?: string | undefined;
//     /**
//      * - extracted field values, placed into an array. This simplifies (at expense of memory) type analysis and summarization when creating the `AggregateSummary`.
//      */
//     value?: AggregateSummary | undefined;
//     /**
//      * - summary of array of string (or decimal) sizes, pre processing into an AggregateSummary
//      */
//     length?: AggregateSummary | undefined;
//     /**
//      * - only applies to Float types. Summary of array of sizes of the value both before and after the decimal.
//      */
//     precision?: AggregateSummary | undefined;
//     /**
//      * - only applies to Float types. Summary of array of sizes of the value after the decimal.
//      */
//     scale?: AggregateSummary | undefined;
//     /**
//      * - if enum rules were triggered will contain the detected unique values.
//      */
//     enum?: string[] | number[] | undefined;
//     /**
//      * - number of times the type was matched
//      */
//     count: number;
//     /**
//      * - absolute priority of the detected TypeName, defined in the object `typeRankings`
//      */
//     rank: number;
// };
// /**
//  * This is an internal intermediate structure.
//  * It mirrors the `FieldSummary` type it will become.
//  */
// export type InternalFieldTypeData = {
//     /**
//      * - array of values, pre processing into an AggregateSummary
//      */
//     value?: any[] | undefined;
//     /**
//      * - array of string (or decimal) sizes, pre processing into an AggregateSummary
//      */
//     length?: number[] | undefined;
//     /**
//      * - only applies to Float types. Array of sizes of the value both before and after the decimal.
//      */
//     precision?: number[] | undefined;
//     /**
//      * - only applies to Float types. Array of sizes of the value after the decimal.
//      */
//     scale?: number[] | undefined;
//     /**
//      * - number of times the type was matched
//      */
//     count?: number | undefined;
//     /**
//      * - absolute priority of the detected TypeName, defined in the object `typeRankings`
//      */
//     rank?: number | undefined;
// };
// /**
//  * Used to represent a number series of any size.
//  * Includes the lowest (`min`), highest (`max`), mean/average (`mean`) and measurements at certain `percentiles`.
//  */
// export type AggregateSummary = {
//     min: number;
//     max: number;
//     mean: number;
//     p25: number;
//     p33: number;
//     p50: number;
//     p66: number;
//     p75: number;
//     p99: number;
// };
// /**
//  * This callback is displayed as a global member.
//  */
// export type progressCallback = (progress: {
//     totalRows: number;
//     currentRow: number;
// }) => any;
// /**
//  * Includes the results of main top-level schema.
//  * @typedef TypeSummary
//  * @type {{
//  *  fields: Object.<string, FieldInfo>;
//  *  totalRows: number;
//  *  nestedTypes?: Object.<string, TypeSummary>;
//  * }}
//  */
// /**
//  * Describes one or more potential types discovered for a field. The `types` object will have a `$ref` key if any nested structures were found.
//  * @typedef FieldInfo
//  * @type {object}
//  * @property {Object.<string, string | FieldTypeSummary>} types - field stats organized by type
//  * @property {boolean} nullable - is the field nullable
//  * @property {boolean} unique - is the field unique
//  * @property {string[]|number[]} [enum] - enumeration detected, the values are listed on this property.
//  */
// /**
//  * Contains stats for a given field's (potential) type.
//  *
//  * TODO: Add string property for the type name.
//  *    We currently uses object key structure: {"String": FieldTypeSummary}
//  * @typedef FieldTypeSummary
//  * @type {Object}
//  * @property {string} [typeAlias] - for nested type support.
//  * @property {AggregateSummary} [value] - extracted field values, placed into an array. This simplifies (at expense of memory) type analysis and summarization when creating the `AggregateSummary`.
//  * @property {AggregateSummary} [length] - summary of array of string (or decimal) sizes, pre processing into an AggregateSummary
//  * @property {AggregateSummary} [precision] - only applies to Float types. Summary of array of sizes of the value both before and after the decimal.
//  * @property {AggregateSummary} [scale] - only applies to Float types. Summary of array of sizes of the value after the decimal.
//  * @property {string[]|number[]} [enum] - if enum rules were triggered will contain the detected unique values.
//  * @property {number} count - number of times the type was matched
//  * @property {number} rank - absolute priority of the detected TypeName, defined in the object `typeRankings`
//  */
// /**
//  * This is an internal intermediate structure.
//  * It mirrors the `FieldSummary` type it will become.
//  * @private
//  * @typedef InternalFieldTypeData
//  * @type {Object}
//  * @property {any[]} [value] - array of values, pre processing into an AggregateSummary
//  * @property {number[]} [length] - array of string (or decimal) sizes, pre processing into an AggregateSummary
//  * @property {number[]} [precision] - only applies to Float types. Array of sizes of the value both before and after the decimal.
//  * @property {number[]} [scale] - only applies to Float types. Array of sizes of the value after the decimal.
//  * @property {number} [count] - number of times the type was matched
//  * @property {number} [rank] - absolute priority of the detected TypeName, defined in the object `typeRankings`
//  *
//  */
// /**
//  * Used to represent a number series of any size.
//  * Includes the lowest (`min`), highest (`max`), mean/average (`mean`) and measurements at certain `percentiles`.
//  * @typedef AggregateSummary
//  * @type {{min: number, max: number, mean: number, p25: number, p33: number, p50: number, p66: number, p75: number, p99: number}}
//  */
// /**
//  * This callback is displayed as a global member.
//  * @callback progressCallback
//  * @param {{totalRows: number, currentRow: number}} progress - The current progress of processing.
//  */
// /**
//  * schemaAnalyzer() is the main function and where all the analysis & processing happens.
//  * @param {string} schemaName The name, or name prefix to use when assembling results. Helpful with nested types (aka sub-types.)
//  * @param {Array<Object>} input - The input data to analyze. Must be an array of objects.
//  * @param {ISchemaAnalyzerOptions} [options] - Optional parameters
//  * @returns {Promise<TypeSummary>} Returns and
//  */
// export function schemaAnalyzer(schemaName: string, input: Array<Object>, options?: ISchemaAnalyzerOptions | undefined): Promise<TypeSummary>;

// export function pivotFieldDataByType(typeSizeData: any): any;
// /**
//  * Accepts an array of numbers and returns summary data about
//  *  the range & spread of points in the set.
//  *
//  * @param {number[]} numbers - sequence of unsorted data points
//  * @returns {AggregateSummary}
//  */
// export function getNumberRangeStats(numbers: number[], useSortedDataForPercentiles?: boolean): AggregateSummary;
// export function isValidDate(date: any): any;
// /**
//  * Returns a fieldName keyed-object with type detection summary data.
//  *
//  * ### Example `fieldSummary`:
//  * ```
//  * {
//  *  "id": {
//  *    "UUID": {
//  *      "rank": 2,
//  *      "count": 25
//  *    },
//  *    "Number": {
//  *      "rank": 8,
//  *      "count": 1,
//  *      "value": {
//  *        "min": 9999999,
//  *        "mean": 9999999,
//  *        "max": 9999999,
//  *        "p25": 9999999,
//  *        "p33": 9999999,
//  *        "p50": 9999999,
//  *        "p66": 9999999,
//  *        "p75": 9999999,
//  *        "p99": 9999999
//  *      }
//  *    }
//  *  }
//  * }
//  * ```
//  *
//  * @param {{ fieldsData: Object.<string, InternalFieldTypeData[]>, uniques: Object.<string, any[]>, totalRows: number}} schema
//  * @returns {{
//  *  fields: Object.<string, Object.<string, FieldTypeSummary>>,
//  *  uniques: Object.<string, any[]>,
//  *  totalRows: number
//  * }}
//  */
// declare function condenseFieldData(schema: {
//     fieldsData: {
//         [x: string]: InternalFieldTypeData[];
//     };
//     uniques: {
//         [x: string]: any[];
//     };
//     totalRows: number;
// }): {
//     fields: {
//         [x: string]: {
//             [x: TypeNameString]: FieldTypeSummary;
//         };
//     };
//     uniques: {
//         [x: string]: any[];
//     };
//     totalRows: number;
// };
// /**
//  *
//  */
// declare function formatRangeStats(stats: any, formatter: any): {
//     min: any;
//     mean: any;
//     max: any;
//     p25: any;
//     p33: any;
//     p50: any;
//     p66: any;
//     p75: any;
//     p99: any;
// };
// export { condenseFieldData as _condenseFieldData, pivotFieldDataByType as _pivotFieldDataByType, getNumberRangeStats as _getNumberRangeStats, formatRangeStats as _formatRangeStats };
