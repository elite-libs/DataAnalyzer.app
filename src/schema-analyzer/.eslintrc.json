{
  "parserOptions": {
    "ecmaVersion": 2020,
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "env": {
    "es6": true,
    "node": true,
    "browser": true,
    "jest": true
  },
  "plugins": ["import", "node", "promise", "prettier"],
  "extends": [
    "prettier",
    "prettier/@typescript-eslint",
    "prettier/babel",
    "prettier/react"
  ]
}
