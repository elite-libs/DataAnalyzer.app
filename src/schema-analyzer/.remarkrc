{
  "settings": {
    "listItemIndent": "mixed"
  },
  "plugins": [
    "remark-stringify",
    ["remark-lint-list-item-indent", false]
  ]
}