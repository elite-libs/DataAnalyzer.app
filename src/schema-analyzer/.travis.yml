language: node_js
node_js:
    - 12
install:
    - npm install -g codecov
script:
    - npm ci
    - npm test
    - codecov
