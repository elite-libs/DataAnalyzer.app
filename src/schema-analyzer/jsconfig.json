{
  "compilerOptions": {
    "jsx": "react",
    "checkJs": true,
    "resolveJsonModule": true,
    "esModuleInterop": true
  },
  "typeAcquisition": {
    "include": [
      "jest"
    ]
  }
}