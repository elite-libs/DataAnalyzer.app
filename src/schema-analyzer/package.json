{
  "name": "schema-analyzer",
  "version": "2.0.0",
  "description": "Heuristic \"column\" type & size analysis w/ enumeration detection.",
  "type": "commonjs",
  "module": "index.tsx",
  "main": "index.tsx",
  "browser": "dist/schema-analyzer.bundle.js",
  "files": [
    "dist",
    "build/index.d.ts",
    "index.ts"
  ],
  "scripts": {
    "prebuild": "rm -rf dist/*",
    "build": "npm run build:webpack && NODE_ENV=production rollup -c && NODE_ENV=development rollup -c && npm run compress",
    "build:webpack": "webpack -c webpack.config.js",
    "coverage": "jest --coverage --maxWorkers 4 && codecov -f coverage/coverage-final.json",
    "postcoverage": "codecov --token=$CODECOV_TOKEN",
    "dev": "rollup -c -w",
    "test": "jest --runTestsByPath --coverage --maxWorkers 4 ./index.test.ts ./utils/type-helpers.test.ts",
    "test:debug": "node --inspect-brk=9823 ./node_modules/.bin/jest --runTestsByPath ./index.test.ts ./utils/type-*.test.ts",
    "test:watch": "jest --watch",
    "pretest": "npm run build",
    "format": "prettier -w *.js index*.ts utils/*.ts",
    "compress": "zopfli -i2000 --gzip ./dist/schema-analyzer*min.js && brotli --keep --quality=11 -- ./dist/schema-analyzer*min.js"
  },
  "engines": {
    "node": "< 14"
  },
  "types": "./index.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/justsml/schema-analyzer.git"
  },
  "keywords": [],
  "author": {
    "name": "Dan Levy",
    "url": "https://danlevy.net/"
  },
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/justsml/schema-analyzer/issues"
  },
  "homepage": "https://github.com/justsml/schema-analyzer#readme",
  "dependencies": {
    "debug": "^4.3.1",
    "ts-node": "^9.1.1"
  },
  "devDependencies": {
    "@babel/core": "~7.8.3",
    "@babel/helper-compilation-targets": "~7.12.5",
    "@babel/plugin-transform-runtime": "~7.8.3",
    "@babel/preset-env": "~7.8.3",
    "@babel/preset-typescript": "^7.12.7",
    "@babel/runtime-corejs3": "~7.8.3",
    "@rollup/plugin-commonjs": "~11.0.1",
    "@rollup/plugin-node-resolve": "~7.0.0",
    "@rollup/plugin-typescript": "8.1.0",
    "@types/debug": "^4.1.5",
    "@types/jest": "~25.1.0",
    "@types/lodash-es": "^4.17.4",
    "babel-eslint": "~10.0.3",
    "babel-jest": "26.6.3",
    "codecov": "3.8.1",
    "core-js": "3.8.2",
    "csv-parse": "~4.8.5",
    "eslint-config-standard": "~14.1.0",
    "eslint-config-standard-react": "~9.2.0",
    "eslint-plugin-import": "2.22.1",
    "eslint-plugin-node": "~11.0.0",
    "eslint-plugin-promise": "~4.2.1",
    "eslint-plugin-react": "7.22.0",
    "eslint-plugin-standard": "~4.0.1",
    "jest": "26.6.3",
    "lodash-es": "^4.17.20",
    "prettier": "^2.2.1",
    "rollup": "~1.30.1",
    "rollup-plugin-node-builtins": "^2.1.2",
    "rollup-plugin-node-globals": "^1.4.0",
    "rollup-plugin-terser": "~5.2.0",
    "standard": "16.0.3",
    "terser-webpack-plugin": "^4.2.3",
    "ts-jest": "^26.4.4",
    "ts-loader": "^8.0.14",
    "tslib": "^2.1.0",
    "typescript": "^4.1.3",
    "webpack": "^5.17.0",
    "webpack-cli": "^4.4.0"
  },
  "prettier": {
    "endOfLine": "lf",
    "trailingComma": "all",
    "tabWidth": 2,
    "semi": false,
    "singleQuote": true
  },
  "jest": {
    "moduleFileExtensions": [
      "ts",
      "js",
      "mjs"
    ],
    "preset": "ts-jest",
    "moduleNameMapper": {
      "^lodash-es$": "<rootDir>/node_modules/lodash/index.js"
    }
  },
  "babel": {
    "plugins": [],
    "presets": [
      [
        "@babel/preset-env",
        {
          "corejs": {
            "version": "3.8.2",
            "proposals": true
          },
          "modules": "auto",
          "useBuiltIns": "usage",
          "targets": {
            "node": 14,
            "browsers": [
              "Chrome >= 60",
              "Safari >= 10.1",
              "iOS >= 10.3",
              "Firefox >= 54",
              "Edge >= 15"
            ]
          }
        }
      ],
      "@babel/preset-typescript"
    ]
  },
  "standard": {
    "env": [
      "jest"
    ]
  },
  "eslintIgnore": [
    "node_modules",
    "dist"
  ],
  "eslintConfig": {
    "parserOptions": {
      "ecmaVersion": 2020,
      "ecmaFeatures": {
        "jsx": true
      }
    },
    "env": {
      "es6": true,
      "node": true,
      "browser": true,
      "jest": true
    },
    "plugins": [
      "import",
      "node",
      "promise",
      "prettier"
    ]
  },
  "husky": {
    "hooks": {
      "pre-commit": "npm run lint",
      "pre-push": "CI=true npm test"
    }
  },
  "browserslist": [
    ">2%",
    "not dead",
    "not ie <= 11",
    "not op_mini all"
  ],
  "remarkConfig": {
    "settings": {
      "listItemIndent": "mixed"
    },
    "plugins": [
      "remark-stringify",
      [
        "remark-lint-list-item-indent",
        false
      ]
    ]
  }
}
