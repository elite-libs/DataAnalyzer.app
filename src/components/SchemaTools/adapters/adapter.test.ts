/* eslint-disable import/first */

import { schemaAnalyzer, helpers } from 'schema-analyzer/index';

const flattenTypes = helpers.flattenTypes;
import knex from './writer.knex';
import users from '../../../../public/users.example.json';
import usersSparse from '../../../../public/user_sparse-subtypes.json';
import people from '../../../../public/swapi-people.json';
import commerceDeptNews from '../../../../public/commerce-dept-news.json';
// import people from '../../../../public/swapi-people.json'
import path from 'path';
import fs from 'fs';
import csvParse from 'csv-parse';

const productCsv: Promise<any[]> = parseCsv(
  fs.readFileSync(
    path.resolve(__dirname, '../../../../public/products-3000.csv'),
    'utf8',
  ),
);

const usersCsv: Promise<any[]> = parseCsv(
  fs.readFileSync(
    path.resolve(__dirname, '../../../../public/users-alt.csv'),
    'utf8',
  ),
);

const flattenWrapper = (result) => {
  result = flattenTypes(result, {
    nullableRowsThreshold: 0.0001,
    targetValue: 'p99',
    targetLength: 'p99',
    targetPrecision: 'p99',
    targetScale: 'p99',
  });
  // if (!isCI) console.log(JSON.stringify(result, null, 2))
  return result;
};

function parseCsv(content): Promise<any[]> {
  return new Promise((resolve, reject) => {
    csvParse(
      content,
      {
        columns: true,
        trim: true,
        skip_empty_lines: true,
      },
      (err, results, info) => {
        if (err) return reject(err);
        resolve(results);
      },
    );
  });
}

describe('knex writer', () => {
  it('can emit correct knex migration', async () => {
    const options = { strictMatching: false };
    const results = await schemaAnalyzer('users', users, options);
    const flatResult = helpers.flattenTypes(results, {
      targetLength: 'p99',
      targetPrecision: 'p99',
      targetScale: 'p99',
      targetValue: 'p99',
      nullableRowsThreshold: 0.001,
    });
    const code = knex.render({
      schemaName: 'test',
      results: flatResult,
      options,
    });

    expect(code).toMatchSnapshot();
  });

  it('can emit knex migration for people json', async () => {
    const options = { strictMatching: false };
    const results = await schemaAnalyzer('people', people, options);
    const flatResult = helpers.flattenTypes(results, {
      targetLength: 'p99',
      targetPrecision: 'p99',
      targetScale: 'p99',
      targetValue: 'p99',
      nullableRowsThreshold: 0.001,
    });
    const code = knex.render({
      schemaName: 'test',
      results: flatResult,
      options,
    });

    expect(code).toMatchSnapshot();
  });

  it('can emit knex migration for nested user json', async () => {
    const options = { strictMatching: false };
    const results = await schemaAnalyzer('users', usersSparse, options);
    const flatResult = helpers.flattenTypes(results, {
      targetLength: 'p99',
      targetPrecision: 'p99',
      targetScale: 'p99',
      targetValue: 'p99',
      nullableRowsThreshold: 0.001,
    });
    const code = knex.render({
      schemaName: 'test',
      results: flatResult,
      options,
    });

    expect(code).toMatchSnapshot();
  });
});
