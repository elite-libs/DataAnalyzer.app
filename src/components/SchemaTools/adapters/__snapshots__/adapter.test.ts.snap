// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#knex can emit migration 1`] = `
"// More info: http://knexjs.org/#Schema-createTable

exports.up = async function up(knex) {
  /* Note: no nested types detected */

  return knex.schema.createTable(\\"test\\", (table) => {
    table.uuid(\\"id\\").notNullable().primary();
    table.string(\\"email\\", 40).notNullable();
    table.string(\\"first_name\\", 20).notNullable();
    table.string(\\"last_name\\", 20).notNullable();
    table.boolean(\\"is_active\\").notNullable();
    table.datetime(\\"created_at\\").notNullable();
    table.datetime(\\"modified_at\\").notNullable();
    table.string(\\"catch_phrase\\", 60).notNullable();
    table.string(\\"company_name\\", 20).notNullable();
    table.integer(\\"credits\\").enum('1', '15', '20', '10', '0', '11', '7', '9').notNullable();
    table.decimal(\\"sales_score\\", 10, 8).notNullable();
    table.string(\\"location\\", 20).enum('nyc', 'remote', 'denver').notNullable();
    table.string(\\"profile\\", 400).unique();
})

};

exports.down = async function down(knex) {
  await knex.schema.dropTableIfExists(\\"test\\");
};
"
`;

exports[`#knex can emit migration for nested type 1`] = `
"// More info: http://knexjs.org/#Schema-createTable

exports.up = async function up(knex) {
  await knex.schema.createTable(\\"users_notes\\", (table) => {
    table.increments(\\"id\\");
    table.string(\\"text\\", 20).notNullable();
})
;


  return knex.schema.createTable(\\"users_sparse\\", (table) => {
    table.increments(\\"id\\");
    table.string(\\"name\\", 20).notNullable();
    table.integer(\\"notes\\").references('id').inTable('users_notes');
})

};

exports.down = async function down(knex) {
  await knex.schema.dropTableIfExists(\\"users_notes\\");
  await knex.schema.dropTableIfExists(\\"users_sparse\\");
};
"
`;

exports[`#knex can emit migration for people json 1`] = `
"// More info: http://knexjs.org/#Schema-createTable

exports.up = async function up(knex) {
  /* Note: no nested types detected */

  return knex.schema.createTable(\\"people\\", (table) => {
    table.string(\\"name\\", 40).notNullable().unique();
    table.integer(\\"height\\");
    table.integer(\\"mass\\");
    table.string(\\"hair_color\\", 20).notNullable();
    table.string(\\"skin_color\\", 20);
    table.string(\\"eye_color\\", 20).enum('blue', 'yellow', 'red', 'brown', 'blue-gray', 'black', 'orange', 'hazel').notNullable();
    table.string(\\"birth_year\\", 20);
    table.string(\\"gender\\", 20).enum('male', 'n/a', 'female', 'hermaphrodite', 'none').notNullable();
    table.string(\\"homeworld\\", 40).notNullable();
    table.json(\\"films\\").notNullable();
    table.json(\\"species\\").notNullable();
    table.json(\\"vehicles\\").notNullable();
    table.json(\\"starships\\").notNullable();
    table.datetime(\\"created\\").notNullable();
    table.datetime(\\"edited\\").notNullable();
    table.string(\\"url\\", 40).notNullable().unique();
    table.increments(\\"id\\");
})

};

exports.down = async function down(knex) {
  await knex.schema.dropTableIfExists(\\"people\\");
};
"
`;

exports[`#knex can emit migration with bigInteger id 1`] = `
"// More info: http://knexjs.org/#Schema-createTable

exports.up = async function up(knex) {
  await knex.schema.createTable(\\"users_notes\\", (table) => {
    table.increments(\\"id\\");
    table.string(\\"text\\", 20).notNullable();
})
;


  return knex.schema.createTable(\\"users_big_int\\", (table) => {
    table.increments(\\"id\\");
    table.string(\\"name\\", 20).notNullable();
    table.integer(\\"notes\\").references('id').inTable('users_notes');
})

};

exports.down = async function down(knex) {
  await knex.schema.dropTableIfExists(\\"users_notes\\");
  await knex.schema.dropTableIfExists(\\"users_big_int\\");
};
"
`;

exports[`#mongoose can emit schema 1`] = `
"const mongoose = require(\\"mongoose\\");
const {Schema, Model, Types} = mongoose;

const Users = new Schema({
  id: {
    type: \\"UUID\\",
    unique: true,
    required: true,
  },
  email: {
    type: \\"String\\",
    required: true,
  },
  firstName: {
    type: \\"String\\",
    required: true,
  },
  lastName: {
    type: \\"String\\",
    required: true,
  },
  isActive: {
    type: \\"Boolean\\",
    required: true,
  },
  createdAt: {
    type: \\"Date\\",
    required: true,
  },
  modifiedAt: {
    type: \\"Date\\",
    required: true,
  },
  catchPhrase: {
    type: \\"String\\",
    required: true,
  },
  companyName: {
    type: \\"String\\",
    required: true,
  },
  credits: {
    type: \\"Number\\",
    required: true,
    enum: [\\"1\\", \\"15\\", \\"20\\", \\"10\\", \\"0\\", \\"11\\", \\"7\\", \\"9\\"],
  },
  salesScore: {
    type: \\"Float\\",
    required: true,
  },
  location: {
    type: \\"String\\",
    required: true,
    enum: [\\"nyc\\", \\"remote\\", \\"denver\\"],
  },
  profile: {
    type: \\"String\\",
    unique: true,
  }});

const usersModel = mongoose.model(\\"users\\", Users);
module.exports.Users = usersModel;
"
`;

exports[`#mongoose can emit schema for people json 1`] = `
"const mongoose = require(\\"mongoose\\");
const {Schema, Model, Types} = mongoose;

const People = new Schema({
  name: {
    type: \\"String\\",
    unique: true,
    required: true,
  },
  height: {
    type: \\"Number\\",
  },
  mass: {
    type: \\"Number\\",
  },
  hairColor: {
    type: \\"String\\",
    required: true,
  },
  skinColor: {
    type: \\"String\\",
  },
  eyeColor: {
    type: \\"String\\",
    required: true,
    enum: [\\"blue\\", \\"yellow\\", \\"red\\", \\"brown\\", \\"blue-gray\\", \\"black\\", \\"orange\\", \\"hazel\\"],
  },
  birthYear: {
    type: \\"String\\",
  },
  gender: {
    type: \\"String\\",
    required: true,
    enum: [\\"male\\", \\"n/a\\", \\"female\\", \\"hermaphrodite\\", \\"none\\"],
  },
  homeworld: {
    type: \\"String\\",
    required: true,
  },
  films: {
    type: \\"Array\\",
    required: true,
  },
  species: {
    type: \\"Array\\",
    required: true,
  },
  vehicles: {
    type: \\"Array\\",
    required: true,
  },
  starships: {
    type: \\"Array\\",
    required: true,
  },
  created: {
    type: \\"Date\\",
    unique: true,
    required: true,
  },
  edited: {
    type: \\"Date\\",
    unique: true,
    required: true,
  },
  url: {
    type: \\"String\\",
    unique: true,
    required: true,
  },
  id: {
    type: \\"Number\\",
    unique: true,
    required: true,
  }});

const peopleModel = mongoose.model(\\"people\\", People);
module.exports.People = peopleModel;
"
`;

exports[`#mongoose can emit schema with nested types 1`] = `
"const mongoose = require(\\"mongoose\\");
const {Schema, Model, Types} = mongoose;

const UsersSparse = new Schema({
  id: {
    type: \\"Number\\",
    unique: true,
    required: true,
  },
  name: {
    type: \\"String\\",
    required: true,
  },
  notes: {
    type: \\"Array\\",
    required: true,
  }});

const usersSparseModel = mongoose.model(\\"usersSparse\\", UsersSparse);
module.exports.UsersSparse = usersSparseModel;
const mongoose = require(\\"mongoose\\");
const {Schema, Model, Types} = mongoose;

const UsersNotes = new Schema({
  id: {
    type: \\"Number\\",
    unique: true,
    required: true,
  },
  text: {
    type: \\"String\\",
    required: true,
  }});

const usersNotesModel = mongoose.model(\\"usersNotes\\", UsersNotes);
module.exports.UsersNotes = usersNotesModel;
"
`;

exports[`#typescript can emit interface for nested types 1`] = `
"export interface UsersSparse {
  id: number;
  name: string;
  notes: UsersNotes[];
};

export interface UsersNotes {
  id: number;
  text: string;
};

"
`;

exports[`#typescript can emit interface for people json 1`] = `
"export interface People {
  name: string;
  height?: number  | null;
  mass?: number  | null;
  hairColor: string;
  skinColor?: string  | null;
  eyeColor: \\"blue\\" | \\"yellow\\" | \\"red\\" | \\"brown\\" | \\"blue-gray\\" | \\"black\\" | \\"orange\\" | \\"hazel\\";
  birthYear?: string  | null;
  gender: \\"male\\" | \\"n/a\\" | \\"female\\" | \\"hermaphrodite\\" | \\"none\\";
  homeworld: string;
  films: any[];
  species: any[];
  vehicles: any[];
  starships: any[];
  created: Date;
  edited: Date;
  url: string;
  id: number;
};

"
`;

exports[`#typescript can emit interface(s) 1`] = `
"export interface Users {
  id: string;
  email: string;
  firstName: string;
  lastName: string;
  isActive: boolean;
  createdAt: Date;
  modifiedAt: Date;
  catchPhrase: string;
  companyName: string;
  credits: \\"1\\" | \\"15\\" | \\"20\\" | \\"10\\" | \\"0\\" | \\"11\\" | \\"7\\" | \\"9\\";
  salesScore: number;
  location: \\"nyc\\" | \\"remote\\" | \\"denver\\";
  profile?: string  | null;
};

"
`;
